### ISSUE TITLE
Using Pretrained NMT

### ISSUE THREAD

**Issue by CheapsiTalk** [bert-nmt/bert-nmt#28]
Hello! Strictly speaking this is not an issue but a question. I have read through your paper, but I do not understand the purpose and effect of using a pretrained NMT model. In the bert-fused model, how much of a difference in model performance is there between using a pretrained NMT model and a randomly initialized one, specifcally in low-resource scenario with unlabeled data?

**Comment by teslacool** [comment]
Actually, we have discussed this problem in our ablation study (1), and you can read this section for more details.

**Comment by CheapsiTalk** [comment]
Thanks for the reply! For the unsupervised bert-fused model, I see your paper said that you are using pretrained XLM model based on same model configurations of the paper "Cross-lingual Language Model Pretraining". How exactly do I simulate that on, say arabic to english translation on my collected dataset? Could you point me to some steps in doing that please? Thank you again!

**Comment by teslacool** [comment]
Yes, we use the [XLM codebase](https://github.com/facebookresearch/XLM) and XLM models released by them. For some reasons, the code has not been organized, and I will use the summer vacation to clean up and release source code as soon as possible.

**Comment by inigo-jauregi-rozetta** [comment]
Hi,

In my case, the model is worse than the pre-trained NMT baseline, even if I use the pre-trained NMT to initialize the models weights. I use the following command:

`CUDA_VISIBLE_DEVICES=0 fairseq-train $preprocess_output --lr 0.0005  -s en -t eu --optimizer adam --max-tokens 4000 --dropout 0.3 --arch transformer_iwslt_de_en --save-dir $training_output --no-epoch-checkpoints --max-epoch 50  --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr '1e-07' --min-lr '1e-09' --adam-betas '(0.9, 0.98)' --weight-decay 0.0001 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --bert-model-name $bert_model --share-all-embeddings --encoder-bert-dropout --encoder-bert-dropout-ratio 0.5 --warmup-from-nmt --reset-lr-scheduler`

It loads the pre-trained model checkpoint correctly, as indicated by the following log:

`Model will load checkpoint from <PATH>
| loaded checkpoint <PATH> (epoch 28 @ updates)
| loading train data for epoch 28`

However, in the first training step, the loss is already much larger, as if the model was trained from zero again. Could you tell me what I am doing wrong? Do I need to lower the learning rate for the training of the BERT-fused model?

**Comment by teslacool** [comment]
It is abnormal, if the initial model seems to be trained from scratch. You should check these two points.
1.  The architecture except bert module is the same as the pretrained one.
2. The vocab size should be the same.

From the logs you provided, I can not see the problem. You can paste these two logs more completely, and I will check more carefully.

**Comment by inigo-jauregi-rozetta** [comment]
Thanks for your comment.

I double-checked what you indicated and noticed that the vocab size wasn't the same in both models. This was due to the fact that I didn't use `--share-all-embeddings` on the pre-trained model.

After removing that argument, the bert-fused model is performing better than the pre-trained NMT model.

**Comment by SefaZeng** [comment]
> Thanks for your comment.
> 
> I double-checked what you indicated and noticed that the vocab size wasn't the same in both models. This was due to the fact that I didn't use `--share-all-embeddings` on the pre-trained model.
> 
> After removing that argument, the bert-fused model is performing better than the pre-trained NMT model.

Hi， @inigo-jauregi-rozetta I meet the same problem like you said the loss is very large, and the bert-fused model get a very bakd performance. Is the `--share-all-embeddings` the core reason?

**Comment by teslacool** [comment]
The `--share-all-embeddings` should not be the reason.
